### ISSUE TITLE
【PaddlePaddle Hackathon 2】10、为 Paddle 新增 AdaptiveLogSoftmaxWithLoss API

### ISSUE THREAD

**Issue by TCChenlong** [PaddlePaddle/Paddle#40325]
（此 ISSUE 为 PaddlePaddle Hackathon 第二期活动的任务 ISSUE，更多详见 [【PaddlePaddle Hackathon 第二期】任务总览](https://github.com/PaddlePaddle/Paddle/issues/40234)）

 【任务说明】

任务标题：为 Paddle 新增 AdaptiveLogSoftmaxWithLoss API

技术标签：深度学习框架，Python

任务难度：中等

详细描述： AdaptiveLogSoftmaxWithLoss 来源于 [Efficient softmax approximation for GPUs](https://arxiv.org/abs/1609.04309)，其中的 Adaptive Softmax 方法是对一种高效实现 softmax 函数近似计算的方法。此任务的目标是在 Paddle 框架中，新增 AdaptiveLogSoftmaxWithLoss API，调用路径为：paddle.nn.AdaptiveLogSoftmaxWithLoss 和 paddle.nn.functional.adaptive_log_softmax_with_loss。

【提交流程】

请按 [新增API 开发&提交流程](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/api_contributing_guides_cn.html)中的描述，完成任务提交过程中的相关步骤。

【提交内容】

- API 的设计文档，并提 PR 至 [PaddlePaddle/community](https://github.com/PaddlePaddle/community) 的 rfcs/APIs 目录
- Python 实现代码 & 英文 API 文档，在 Paddle repo 的 [python/paddle/nn/layer/loss.py](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/nn/layer/loss.py) 文件 和 [python/paddle/nn/functional/loss.py](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/nn/functional/loss.py) 文件
- 单测代码，在 Paddle repo 的 [python/paddle/fluid/tests/unittests](https://github.com/PaddlePaddle/Paddle/tree/develop/python/paddle/fluid/tests/unittests) 目录
- 中文API文档，在 docs repo 的 [docs/api/paddle/nn](https://github.com/PaddlePaddle/docs/tree/develop/docs/api/paddle/nn) 目录和 [docs/api/paddle/nn/functional/](https://github.com/PaddlePaddle/docs/tree/develop/docs/api/paddle/nn/functional)

【合入标准】

- 1、按 [API 设计规范](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/api_design_guidelines_standard_cn.html) 完成 API设计文档；
- 2、按[API 验收标准](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/api_accpetance_criteria_cn.html)完成 API功能实现、单测、API文档；

【技术要求】

- 熟悉 Adaptive Softmax 算法原理和适用场景
- 熟练掌握 Python

【参考内容】

- [文档：API 贡献指南](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/api_contributing_guides_cn.html)

【答疑交流】

- 如果在开发中对于上述任务有任何问题，欢迎在本 ISSUE 下留言交流。
- 对于开发中的共性问题，在活动过程中，会定期组织答疑，请大家关注官网&QQ群的通知，及时参与。

**Comment by renmada** [comment]
API设计文档 有模板吗

**Comment by thunder95** [comment]
AdaptiveLogSoftmaxWithLoss在torch中是一个组网的layer层， 里面里有好几个可学习的Linear，如是要在[python/paddle/nn/functional/loss.py](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/nn/functional/loss.py) 这里面实现，是需要每次把weight和bias都传进去吗？

**Comment by TCChenlong** [comment]
> API设计文档 有模板吗

[community/api_design_template.md at master · PaddlePaddle/community](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/api_design_template.md) 这个就是哈~
